lib/upload_to_s3.js

```diff
@@ -1,45 +1,81 @@
 import https from 'https';
 import url from 'url';
+import fs from 'fs';
+import temporary from 'temporary';
+import promiseRetry from 'promise-retry';
 
 export default async function uploadToS3 (task, sourceStream, artifactName, expiration, httpsHeaders) {
-  var queue = task.runtime.queue;
-
-  var artifact = await queue.createArtifact(
-    task.status.taskId,
-    task.runId,
-    artifactName,
-    {
-      // Why s3? It's currently cheaper to store data in s3 this could easily
-      // be used with azure simply by changing s3 -> azure.
-      storageType: 's3',
-      expires: new Date(Math.min(expiration, new Date(task.task.expires))),
-      contentType: httpsHeaders['content-type']
-    }
-  );
-
-  let parsedUrl = url.parse(artifact.putUrl);
-  let options = {
-    hostname: parsedUrl.hostname,
-    path: parsedUrl.path,
-    method: 'PUT',
-    headers: httpsHeaders
-  };
-
-  let req = https.request(options);
-  sourceStream.pipe(req);
-
-  let response = await new Promise((accept, reject) => {
-    req.on('response', accept);
-    req.on('error', reject);
-    req.setTimeout(5 * 60 * 1000, reject);
+  let queue = task.runtime.queue;
+
+  var artifact;
+  var parsedUrl;
+  var options;
+
+  let tmp = new temporary.File();
+  await new Promise((accept, reject) => {
+    let stream = fs.createWriteStream(tmp.path);
```

Comment:
I think that this method should only be responsible for uploading a file, not also storing it somewhere.  We could wrap docker.copy() to save to a temp file, and then here we'll open up a read stream from it.